packages/connection/src/class-plugin-storage.php

```diff
@@ -0,0 +1,97 @@
+<?php
+/**
+ * Storage for plugin connection information.
+ *
+ * @package automattic/jetpack-connection
+ */
+
+namespace Automattic\Jetpack\Connection;
+
+/**
+ * The class serves a single purpose - to store the data that plugins use the connection, along with some auxiliary information.
+ * Well, we don't really store all that. The information is provided on runtime,
+ * so all we need to do is to save the data into the class property and retrieve it from there on demand.
+ *
+ * @todo Adapt for multisite installations.
+ */
+class Plugin_Storage {
+
+	const OPTION_KEY = 'connection_plugins';
+
+	/**
+	 * Connected plugins.
+	 *
+	 * @var array
+	 */
+	private static $plugins = array();
+
+	/**
+	 * Whether the plugins were configured.
+	 * To make sure we don't call the configuration process again and again.
+	 *
+	 * @var bool
+	 */
+	private static $plugins_configuration_finished = false;
+
+	/**
+	 * Add or update the plugin information in the storage.
+	 *
+	 * @param string $slug Plugin slug.
+	 * @param array  $args Plugin arguments, optional.
+	 *
+	 * @return bool
+	 */
+	public static function upsert( $slug, array $args = array() ) {
+		self::$plugins[ $slug ] = $args;
+
+		return true;
+	}
+
+	/**
+	 * Retrieve the plugin information by slug.
+	 *
+	 * @param string $slug The plugin slug.
+	 *
+	 * @return array|null
+	 */
+	public static function get_one( $slug ) {
+		return empty( self::$plugins[ $slug ] ) ? null : self::$plugins[ $slug ];
+	}
+
+	/**
+	 * Retrieve info for all plugins that use the connection.
+	 *
+	 * @return array
+	 */
+	public static function get_all() {
+		if ( ! self::$plugins_configuration_finished ) {
+			/**
+			 * Fires upon retrieval of the connected plugins.
+			 * Only fires once, as the data isn't supposed to change after it's been initialized.
```

Comment:
That's a really good point.
If some code calls `get_all()` before the `jetpack-config` has initialized (`plugins_loaded` priority 2), the list of initialized plugins will remain empty.
I have an idea on how to not allow `get_all()` to be called too early. I'll push the fix later today and would love to hear your opinion on that.

Reply:
Hi @kbrown9.
I've pushed the fix, please review: https://github.com/Automattic/jetpack/pull/15456/commits/801a4ee9ed9c50135674d1eedd7007bf82de6f42

Reply:
When `get_all()` is called, `$plugins_configuration_finished` is set to true, and we don't allow any changes to the connected plugins list. If another plugin's connection status changes, then subsequent calls to `get_all()` on the page would not return the latest information, right?

Reply:
By adding the flag `$plugins_configuration_finished` we're only making sure that the action `jetpack_connection_configure_plugin` doesn't get run over and over again. The action is only used by `Config` to pass through information on the plugins (name, slug, some other info), it doesn't indicate whether the connection is active. This is kind of lazy initialization.

After the initialization, all the changes are applied directly to the `Plugin_Storage::$plugins`. If a plugin gets removed, it can call `$manager->get_plugin()->remove()` to remove its info from the list. However, I can't think of a scenario where this can be needed, since the page will get reloaded after that happens.

We can run into issues if other developers use `jetpack_connection_configure_plugin` incorrectly, but that's the standard behavior of WordPress hooks - many of them are only called once, so if you're adding an action for a hook that's already been run, you're too late for the party.

Either way, I'd love to hear your thoughts on the solution as a whole, because I do realize that this `$plugin_configuration_finished` approach seems to be overly complex at times.

Reply:
(This is a review as well as me getting familiar with everything)

Looking at this, it feels a bit out of a place to register the plugins in a hook in the `get` method. Get method should get registered plugins, not register anything.

Since this is a pure static class, I wonder if we need this hook. What if the code in `Config' simply called `Plugin->add() ` directly, instead of hooking it? By that time, we know the Plugin_Storage class is loaded, so this should be enough... Makes sense?

Reply:
That's sort of a lazy loading for the connected plugins list.
It's not very likely that we'll need that information (it's only used during disconnection process), so we're trying to not run that initialization unless something actually requests it.

It doesn't play a huge role on the performance side right now, as we're only adding the plugins data into the array, but as we go further it will play more significant role.

Besides, it provides a convinient way for plugins to initialize the connection without `Jetpack\Config`. I'm not sure if we should take that into account though.

Reply:
Another thing. If we call `get_one()` before any call to `get_all()` we won't find what we're looking for

Reply:
Great catch, I really missed that!
Fixed now, thanks!